tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+
+    if (installPath.startsWith("..")) {
+      // absPath is not a subdirectory of this.sourceRoot.
+      return;
+    }
+
+    const dirs = this._splitPath(pathDirname(installPath));
+    const bundlingClientApp =
+      ! this.name && // Indicates we are bundling an app.
+      archMatches(this.bundleArch, "web");
+
+    for (let dir of dirs) {
+      if (dir.charAt(0) === "." ||
+          dir === "packages" ||
+          dir === "programs" ||
+          dir === "cordova-build-override") {
+        // These directories are never loaded as part of an app.
+        return;
+      }
+
+      if (bundlingClientApp && (dir === "server" ||
+                                dir === "private")) {
+        // If we're bundling an app for a client architecture, any files
+        // contained by a server-only directory that is not contained by
+        // a node_modules directory must be ignored.
+        return;
+      }
+
+      if (dir === "node_modules") {
+        if (! this._hasKnownExtension(installPath)) {
+          // Reject any files within node_modules directories that do
+          // not have one of the known extensions.
+          return;
+        }
+
+        // Accept any file within a node_modules directory if it has a
+        // known file extension.
+        return {
+          path: installPath,
+          inNodeModules: true,
+        };
+      }
+    }
+
+    return {
+      path: installPath,
+      inNodeModules: false,
+    };
+  }
+
+  _hasKnownExtension(path) {
+    return this.extensions.indexOf(pathExtname(path)) >= 0;
+  }
+
+  _splitPath(path) {
+    const partsInReverse = [];
+    for (let dir; (dir = pathDirname(path)) !== path; path = dir) {
+      partsInReverse.push(pathBasename(path));
+    }
+    return partsInReverse.reverse();
+  }
+
+  _tryToResolveImportedPath(id, path) {
+    let resolved =
+      this._resolveAbsolute(id) ||
+      this._resolveRelative(id, path) ||
+      this._resolveNodeModule(id, path);
+
+    while (resolved && resolved.stat.isDirectory()) {
+      resolved = this._resolvePkgJsonMain(resolved.path) ||
+        this._joinAndStat(dirPath, "index.js");
+    }
+
+    return resolved && resolved.path;
+  }
+
+  _joinAndStat(...joinArgs) {
+    const path = pathNormalize(pathJoin(...joinArgs));
+    const exactStat = statOrNull(path);
+    const exactResult = exactStat && { path, stat: exactStat };
+    if (exactResult && exactStat.isFile()) {
+      return exactResult;
+    }
+
+    for (let ext of this.extensions) {
+      const pathWithExt = path + ext;
+      const stat = statOrNull(pathWithExt);
+      if (stat) {
+        return { path: pathWithExt, stat };
+      }
+    }
+
+    if (exactResult && exactStat.isDirectory()) {
+      // After trying all available file extensions, fall back to the
+      // original result if it was a directory.
+      return exactResult;
+    }
+
+    return null;
+  }
+
+  _resolveAbsolute(id) {
```

Comment:
I'd suggest disallowing absolute paths, then, as you suggest (rather than making them do something distinct from what Node usually does), to help the potential transition towards "just Node projects" instead of creating some new behavior to preserve.